src/_pytest/main.py

```diff
@@ -540,10 +547,22 @@ def pytest_runtest_logreport(
     def isinitpath(self, path: Union[str, "os.PathLike[str]"]) -> bool:
         return Path(path) in self._initialpaths
 
-    def gethookproxy(self, fspath: "os.PathLike[str]"):
+    def gethookproxy(self, fspath: "os.PathLike[str]") -> PathAwareHookProxy:
         # Check if we have the common case of running
         # hooks with all conftest.py files.
+
         pm = self.config.pluginmanager
+        len_conftest_plugins = len(pm._conftest_plugins)
+
+        if len_conftest_plugins != self._num_conftest_plugins_collected:
+            self._fspath_hookproxy_cache.clear()
+            self._num_conftest_plugins_collected = len_conftest_plugins
+
+        proxy: Optional[PathAwareHookProxy] = self._fspath_hookproxy_cache.get(fspath)
+
+        if proxy:
```

Comment:
for the canonical type guard
```suggestion
        if proxy is not None:
```